Add a mixing length turbulence model for INSFV momentum equations

smharper opened the pull request:
Closes #16794.

This PR adds an `INSFVMixingLengthReynoldsStress` kernel and a `WallDistanceMixngLengthAux` auxkernel to the Navier-Stokes module. These kernels give us new capabilities for modeling turbulent fluids. The gritty details of the physics are discussed in a new theory document. There is also a simple circular pipe demo included which shows how we can tune a parameter in the auxkernel to match friction factor correlations.

Here are example velocity profiles from the pipe example. We can see the characteristic log-law type of shape developing which is what we expect from turbulent flow.

![turb_vel_profile](https://user-images.githubusercontent.com/4622797/105550852-25588800-5cbf-11eb-8187-3564c90925fd.png)

Note that the mixing length is implemented as an auxvariable here. I think this will give us some flexibility to try out more detailed models that go beyond assuming the mixing length is simply proportional to the wall distance. This will also be helpful because we can add passive scalar diffusion kernels that use the same mixing length variable.

smharper committed:
Add a mixing length model for Reynolds stress

smharper committed:
Don't use Laplacian form of strain rate for RANS

smharper committed:
Use all velocity components for Reynolds stress

smharper committed:
Allow multiple walls for the mixing length model

smharper committed:
INSFVRANSForce -> INSFVMixingLengthReynoldsStress

The new name, INSFVMixingLengthReynoldsStress, makes the dependence on a mixing
length model explicit

smharper committed:
Add mixing length syntax docs

smharper committed:
Add a mixing length test

smharper committed:
Add mixing length theory documentation

smharper committed:
Add a mixing length tuning example

smharper committed:
Cleanup theory docs; reference a MOOSE issue

Refs #16794

loganharbour commented:
In an ideal world - you should reference the issue in more than just one commit if possible :-)

smharper committed:
clang-format

smharper committed:
Protect global AD code with ifdefs

smharper committed:
Address PR #16795 issues

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/667691/) on b5e6c5f wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/16795/site)

This comment will be updated on new commits.

lindsayad reviewed (commented):
There are way too much documentation, comments, and references for this to be MOOSE code. I'll have to reject :laughing: I'm just kidding. I think I should be able to review this tomorrow.

lindsayad commented on modules/navier_stokes/examples/pipe_mixing_length/pipe_mixing_length.i:
```diff
@@ -0,0 +1,260 @@
+# This example demonstrates how the mixing length model can be tuned to match an
```
It has been my personal experience that untested input files inevitably stop working over time. What we often do for big examples is a `check_input` test that looks like this:
```
[Tests]
  [pipe_mixing]
    type = RunApp
    input = 'pipe_mixing_length.i'
    check_input = True
  []
[]
```
where we can at least be confident that all the syntax in the input file still works, making it more likely that this example will run two years from now when a user wants to try out the example. I can't recall whether `check_input` actually checks for the existence of the mesh file...it probably does. If so, and the `.msh` file we would need is huge, then perhaps we can comment out the `FileMeshGenerator` and put in a simpler stand-in?

smharper replied:
The `.msh` file weighs 500 kB so it's probably bigger than what we want to commit into the repo. I'd prefer that we don't comment out the real mesh; I'm worried that users would overlook it in the input file. Do you think we could modify the testing infrastructure so we could have something like `gmsh_input = mesh/pipe.geo` to have the test harness itself run gmsh?

lindsayad replied:
The latter would probably be easier and more sustainable. Otherwise we will have to make sure that gmsh is available on our testing boxes

lindsayad replied:
Actually I'm not sure...I don't know how much `check_input` does. For example does it want to do kernel and material integrity checks? If so, then we kind of need the mesh. I see that spack does actually have gmsh so maybe what you're asking for is not impossible

smharper replied:
Turns out that `GeneratedMeshGenerator` includes a `bias_y` parameter! With that we don't need gmsh at all for this example.

lindsayad reviewed (changes requested):
Really great work. Pretty simply requested changes

smharper committed:
Address PR #16795 feedback

smharper commented on modules/navier_stokes/doc/content/modules/navier_stokes/rans_theory.md:
```diff
@@ -0,0 +1,307 @@
+decomposed as $\bm{v} = \overline{\bm{v}} + \bm{v}'$ where $\overline{\bm{v}}$
+is the average value and $\bm{v}'$ is the fluctuating component. It is difficult
+to rigorously define what is meant by ``average'' and what is meant by
+``fluctuating'', but these concepts can be roughly defined as: the average
```
Whoops. I'm writing these files in vim and just did the latex style tics from muscle memory. Fixed now

GiudGiud commented on modules/navier_stokes/test/tests/finite_volume/ins/channel-flow/2d-mixing-length.i:
```diff
@@ -0,0 +1,199 @@
+  type = Steady
+  solve_type = 'NEWTON'
+  petsc_options_iname = '-pc_type -ksp_gmres_restart -sub_pc_type -sub_pc_factor_shift_type'
+  petsc_options_value = 'asm      200                lu           NONZERO'
```
you might run into the same issue I ran into with those, see #16911

smharper replied:
What was the issue? Were the results not reproducible on CIVET with ASM?

GiudGiud replied:
yup, was getting failures in next. Had to use lu

lindsayad replied:
Added the `Navier Stokes` testing target here

smharper replied:
:( yup, got a fail there. Do yall know what the root cause is yet?

GiudGiud replied:
Under investigation in #16911

lindsayad replied:
@smharper your failure is a hard failure unrelated to the work in #16911

lindsayad commented on modules/navier_stokes/src/auxkernels/WallDistanceMixingLengthAux.C:
```diff
@@ -0,0 +1,65 @@
+  const MeshBase & l_mesh = m_mesh.getMesh();
+
+  // Get the ids of the wall boundaries
+  std::vector<BoundaryID> vec_ids = m_mesh.getBoundaryIDs(_wall_boundary_names, true);
```
```suggestion
  std::vector<BoundaryID> vec_ids = m_mesh.getBoundaryIDs(_wall_boundary_names, false);
  std::set<BoundaryID> set_ids(vec_ids.begin(), vec_ids.end());
  _communicator.set_union(set_ids);
```

lindsayad commented on modules/navier_stokes/src/auxkernels/WallDistanceMixingLengthAux.C:
```diff
@@ -0,0 +1,65 @@
+  // Loop over all boundaries
+  Real min_sq_dist = 1e9;
+  const auto & bnd_to_elem_map = m_mesh.getBoundariesToElems();
+  for (BoundaryID bid : vec_ids)
```
```suggestion
  for (BoundaryID bid : set_ids)
```

lindsayad reviewed (changes requested):
@smharper this should fix your distributed mesh failures. `getBoundaryIDs` and the underlying libmesh `BoundaryInfo::get_boundary_ids` only returns semi-local information, so you need the `set_union` in order to globalize.

Thinking about this more...you could probably do a fair amount of the work you're doing in `computeValue` within your aux kernel's constructor and save a fair amount of work. For example everything in front of your `mooseError` is sanity checking that you could do up front. Especially now that I am suggesting we do a `_communicator.set_union`. We do not want to do a global communication for every `computeValue` call. So yea I suggest you just save off the `getBoundariesToElems()` map in a const reference data member for later use in `computeValue`.

smharper commented:
Dang, I hadn't thought about distributed mesh! But, I'm going to have a more fundamental problem than just getting the IDs right? Like we need to think about "What if the nearest boundary element is on a different process?".

lindsayad commented:
Yea you're right about that. It may be time for us to add a relationship manager that simply ghosts all boundary elements

lindsayad commented:
That would not be a difficult one to add

lindsayad commented:
Do you want to take a shot at that? Or would you like me to work on it, and then I can push here to your branch? Or actually the relationship manager itself could be a standalone framework PR

smharper commented:
I'd be worried that some distributed meshes might have pretty fat boundary sets. It's easy to imagine a PWR mesh for example with a sideset for outside of each fuel pellet and inside of each cladding rod. Maybe this AuxKernel should actually be a user object so that it can do all the necessary parallel communication?

smharper commented:
We could essentially repeat all of the logic in `NearstNodeMulitAppTransfer`

lindsayad commented:
That's a fine suggestion as well. I would still hope that almost all of the intense work could be done during `init`

smharper commented:
This is also starting to sounds like a bit of a rabbit hole. I'm leaning towards just adding a `mooseError` saying that only serial meshes are allowed.

lindsayad commented:
Sounds good to me

smharper committed:
Limit wall distance auxkernel to replicated meshes

refs #16937

smharper committed:
Use meshgenerators for the mixing length example

This also allows us to add a "tests" file that checks for input syntax regressions

lindsayad reviewed (approved):
LGTM. I still think you could do some of the work in `computeValue` up front somewhere, but the code would be less clean. If it shows up in profiling, which it probably won't, then maybe we can do something about it.
